File: src/main/java/net/ins/edu/algorithms/leetcode/array/MergeSort.java
package net.ins.edu.algorithms.leetcode.array;

import java.util.Arrays;

public class MergeSort {

    public void mergeSort(int[] a, int len) {
        if (len <= 1) return;

        int mid = a.length / 2;

        int[] l = new int[mid];
        int[] r = new int[len - mid];

        int k = 0;
        for (int i = 0; i < a.length; i++) {
            if (i < mid) {
                l[i] = a[i];
            } else {
                r[k++] = a[i];
            }
        }

        mergeSort(l, mid);
        mergeSort(r, len - mid);

        merge(l, r, a, mid, len - mid);
    }

    private void merge(int[] left, int[] right, int[] source, int leftSize, int rightSize) {
        int i = 0, j = 0, k = 0;

        while (i < leftSize && j < rightSize) {
            if (left[i] < right[j]) {
                source[k++] = left[i++];
            } else {
                source[k++] = right[j++];
            }
        }

        while (i < leftSize) {
            source[k++] = left[i++];
        }

        while (j < rightSize) {
            source[k++] = right[j++];
        }
    }


    public static void main(String[] args) {
        var arr = new int[] { 9, 1, 8, 2, 7, 3, 6, 4, 5, 10 };
        new MergeSort().mergeSort(arr, arr.length);
        System.out.println(Arrays.toString(arr));
    }
}

File: python/search/merge_sort.py
import array


def mergeSort(a: array, size: int) -> None:
    if (size <= 1):
        return

    mid = int(size / 2)
    l = a[:mid]
    r = a[mid:]

    mergeSort(l, len(l))
    mergeSort(r, len(r))

    merge(a, l, r, len(l), len(r))

def merge(a: array, left: array, right: array, l_size: int, r_size: int) -> None:
    l = r = i = 0
    while l < l_size and r < r_size:
        if left[l] < right[r]:
            a[i] = left[l]
            l += 1
        else:
            a[i] = right[r]
            r += 1
        i += 1

    while l < l_size:
        a[i] = left[l]
        i += 1
        l += 1

    while r < r_size:
        a[i] = right[r]
        i += 1
        r += 1
    

if __name__ == '__main__':
    arr: array = [1, 9, 2, 8, 3, 7, 4, 6, 5, 0, 10]
    mergeSort(arr, len(arr))
    print(arr)